packages/jupyter-chat/src/model.ts

```diff
@@ -520,6 +553,53 @@ export class ChatModel implements IChatModel {
    */
   inputChanged?(input?: string): void {}
 
+  /**
+   * Add attachment to send with next message.
+   */
+  addAttachment = (attachment: IAttachment, emit: boolean = true): void => {
+    const index = this.inputAttachments.findIndex(
+      att => att.type === attachment.type && att.value === attachment.value
+    );
+    if (index > -1) {
+      this.inputAttachments.splice(index, 1, attachment);
+    } else {
+      this.inputAttachments.push(attachment);
+    }
+    if (emit) {
+      this._inputAttachmentsChanges.emit([...this.inputAttachments]);
+    }
+  };
+
+  /**
+   * Remove attachment to be sent.
+   */
+  removeAttachment = (attachment: IAttachment): void => {
+    const index = this.inputAttachments.findIndex(
+      att => att.type === attachment.type && att.value === attachment.value
+    );
+    if (index > -1) {
+      this.inputAttachments.splice(index, 1);
+    }
+    this._inputAttachmentsChanges.emit([...this.inputAttachments]);
+  };
```

Comment:
I think that similarly here, it makes sense to only emit the signal when the list changes:


```suggestion
  removeAttachment = (attachment: IAttachment): void => {
    const attachmentIndex = this.inputAttachments.findIndex(
      att => att.type === attachment.type && att.value === attachment.value
    );
    if (attachmentIndex === -1) {
      return;
    }
    
    this.inputAttachments.splice(attachmentIndex, 1);
    this._inputAttachmentsChanges.emit([...this.inputAttachments]);
  };
```